drivers/usb/usbuhci_new/roothub.c

```diff
@@ -0,0 +1,470 @@
+                            IN USHORT Port)
+{
+    DPRINT("UhciRHClearFeaturePortPower: UNIMPLEMENTED. FIXME\n");
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHClearFeaturePortSuspend(IN PVOID uhciExtension,
+                              IN USHORT Port)
+{
+    DPRINT("UhciRHClearFeaturePortSuspend: UNIMPLEMENTED. FIXME\n");
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHClearFeaturePortEnableChange(IN PVOID uhciExtension,
+                                   IN USHORT Port)
+{
+    PUHCI_EXTENSION UhciExtension = uhciExtension;
+    PUHCI_HW_REGISTERS BaseRegister;
+    PUSHORT PortControlRegister;
+    UHCI_PORT_STATUS_CONTROL PortControl;
+
+    DPRINT("UhciRHClearFeaturePortEnableChange: ...\n");
+
+    BaseRegister = UhciExtension->BaseRegister;
+    PortControlRegister = (PUSHORT)&BaseRegister->PortControl[Port - 1];
+    PortControl.AsUSHORT = READ_PORT_USHORT(PortControlRegister);
+
+    PortControl.ConnectStatusChange = FALSE;
+    PortControl.PortEnableDisableChange = TRUE;
+    WRITE_PORT_USHORT(PortControlRegister, PortControl.AsUSHORT);
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHClearFeaturePortConnectChange(IN PVOID uhciExtension,
+                                    IN USHORT Port)
+{
+    PUHCI_EXTENSION UhciExtension = uhciExtension;
+    PUHCI_HW_REGISTERS BaseRegister;
+    PUSHORT PortControlRegister;
+    UHCI_PORT_STATUS_CONTROL PortControl;
+
+    DPRINT("UhciRHClearFeaturePortConnectChange: Port - %04X\n", Port);
+
+    BaseRegister = UhciExtension->BaseRegister;
+    PortControlRegister = (PUSHORT)&BaseRegister->PortControl[Port - 1];
+    PortControl.AsUSHORT = READ_PORT_USHORT(PortControlRegister);
+
+    if (PortControl.ConnectStatusChange == TRUE)
+    {
+        /* WC (Write Clear) bits */
+        PortControl.PortEnableDisableChange = FALSE;
+        PortControl.ConnectStatusChange = TRUE;
+        WRITE_PORT_USHORT(PortControlRegister, PortControl.AsUSHORT);
+    }
+
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHClearFeaturePortResetChange(IN PVOID uhciExtension,
+                                  IN USHORT Port)
+{
+    PUHCI_EXTENSION UhciExtension = uhciExtension;
+    DPRINT("UhciRHClearFeaturePortResetChange: ...\n");
+    UhciExtension->ResetChangePortMask &= ~(1 << (Port - 1));
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHClearFeaturePortSuspendChange(IN PVOID uhciExtension,
+                                    IN USHORT Port)
+{
+    DPRINT("UhciRHClearFeaturePortSuspendChange: UNIMPLEMENTED. FIXME\n");
+    return MP_STATUS_SUCCESS;
+}
+
+MPSTATUS
+NTAPI
+UhciRHClearFeaturePortOvercurrentChange(IN PVOID uhciExtension,
+                                        IN USHORT Port)
+{
+    DPRINT("UhciRHClearFeaturePortOvercurrentChange: UNIMPLEMENTED. FIXME\n");
+    return MP_STATUS_SUCCESS;
+}
+
+VOID
+NTAPI
+UhciRHDisableIrq(IN PVOID uhciExtension)
+{
+    //DPRINT("UhciRHDisableIrq: \n");
```

Comment:
I thought these were pretty important. Is UHCI special in that it doesn't need them?

Reply:
Deleted.